app/lib/service/openid/jwt.dart

```diff
@@ -123,7 +114,179 @@ Map<String, dynamic> _decodePart(String part) {
   return _jsonUtf8Base64.decode(base64.normalize(part)) as Map<String, dynamic>;
 }
 
-DateTime? _tryParseFromSeconds(int? value) {
+DateTime? _parseSecondsIfNotNull(int? value) {
   if (value == null) return null;
-  return DateTime.fromMillisecondsSinceEpoch(value * 1000);
+  return DateTime.fromMillisecondsSinceEpoch(value * 1000, isUtc: true);
+}
+
+/// The parsed JWT header.
+class JwtHeader extends UnmodifiableMapView<String, dynamic> {
+  /// algorithm
+  final String? alg;
+
+  /// type
+  final String? typ;
+
+  /// key identifier
+  final String? kid;
+
+  JwtHeader._(super.map)
+      : alg = map['alg'] as String?,
+        typ = map['typ'] as String?,
+        kid = map['kid'] as String?;
+
+  factory JwtHeader(Map<String, dynamic> values) {
+    try {
+      return JwtHeader._(values);
+    } on FormatException {
+      rethrow;
+    } catch (_) {
+      throw FormatException('Unexpected value in JWT header.');
+    }
+  }
+}
+
+/// The parsed JWT payload.
+class JwtPayload extends UnmodifiableMapView<String, dynamic> {
+  /// timestamp when token was issued
+  final DateTime? iat;
+
+  /// timestamp before which the token is not valid
+  final DateTime? nbf;
+
+  /// timestamp at which the token expires
+  final DateTime? exp;
+
+  JwtPayload._(super.map)
+      : iat = _parseSecondsIfNotNull(map['iat'] as int?),
+        nbf = _parseSecondsIfNotNull(map['nbf'] as int?),
+        exp = _parseSecondsIfNotNull(map['exp'] as int?);
+
+  factory JwtPayload(Map<String, dynamic> map) {
+    try {
+      return JwtPayload._(map);
+    } on FormatException {
+      rethrow;
+    } catch (_) {
+      throw FormatException('Unexpected value in JWT payload.');
+    }
+  }
+
+  /// Returns the first key from [keys] which is missing or has `null` value.
+  ///
+  /// Returns `null` if all the [keys] are present with non-null values.
+  String? firstMissingKey(Iterable<String> keys) {
+    for (final key in keys) {
+      final value = this[key];
+      if (value == null) {
+        return key;
+      }
+    }
+    return null;
+  }
+
+  /// Verifies the timestamps with to the provided [now]:
+  /// - [iat] <= [now]
+  /// - [nbf] <= [now]
+  /// - [now] <= [exp]
+  bool verifyTimestamps([DateTime? now]) {
+    now ??= clock.now();
+    if (iat != null && iat!.isAfter(now)) {
+      return false;
+    }
+    if (nbf != null && nbf!.isAfter(now)) {
+      return false;
+    }
+    if (exp != null && exp!.isBefore(now)) {
+      return false;
+    }
+    return true;
+  }
+}
+
+/// Parsed payload with the payload values GitHub sends with the token.
+class GitHubJwtPayload extends JwtPayload {
```

Comment:
On topic: I don't know if we need this abstraction?

Are we going to make `JsonWebToken` type parameterized?

What are we going to return from the `requireAuthentication` method? when we don't return a user?

@isoos  Let's chat when you have time later this week.

Reply:
`requireAuthentication` needs to return either an authenticated `User` or a service account description, as outlined (but not finalized) in https://github.com/dart-lang/pub-dev/pull/5852.

I'm not yet sure about this abstraction, but it feels like we should use something typed and verified over the plain payload data map.

Reply:
Maybe we merge this into `AuthenticatedGithubAction` so github related parts live there...